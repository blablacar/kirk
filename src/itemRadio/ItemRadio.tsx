import React, { Component, Fragment } from 'react'
import cc from 'classcat'

import { OnChangeParameters } from '../_internals/onChange'
import { RadioIcon } from '../_internals/radioIcon'
import { A11yProps, pickA11yProps } from '../_utils/interfaces'
import { NormalizeProps } from '../layout/layoutNormalizer'
import { TextDisplayType } from '../text'
import { StyledItemRadio } from './ItemRadio.style'

export enum ItemRadioStatus {
  DEFAULT = 'default',
  LOADING = 'loading',
}

export type ItemRadioProps = A11yProps &
  NormalizeProps &
  Readonly<{
    className?: string
    name: string
    value: string | number
    leftAddon?: React.ReactNode
    labelTitle?: string
    label: string
    data?: string
    dataInfo?: string
    checked?: boolean
    disabled?: boolean
    chevron?: boolean
    highlighted?: boolean
    onChange?: (obj: OnChangeParameters) => void
    onClick?: (obj: OnChangeParameters) => void
    status?: ItemRadioStatus
    key?: string | number
  }>

type ItemRadioState = {
  focus: boolean
}

export class ItemRadio extends Component<ItemRadioProps> {
  static defaultProps: Partial<ItemRadioProps> = {
    onChange() {},
    onClick() {},
    checked: false,
  }

  state: ItemRadioState = {
    focus: false,
  }

  onChange = () => {
    const { name, value } = this.props
    this.props.onChange({ name, value })
  }

  onClick = () => {
    const { name, value } = this.props
    this.props.onClick({ name, value })
  }

  onFocus = () => {
    this.setState({ focus: true })
  }

  onBlur = () => {
    this.setState({ focus: false })
  }

  render() {
    const {
      label,
      name,
      value,
      data,
      className,
      labelTitle,
      dataInfo,
      checked,
      disabled,
      status,
      chevron,
      highlighted,
      leftAddon,
      hasHorizontalSpacing = false,
    } = this.props
    const a11yAttrs = pickA11yProps<ItemRadioProps>(this.props)
    const isLoading = status === ItemRadioStatus.LOADING
    const radio = (
      <Fragment>
        <input
          type="radio"
          name={name}
          value={value}
          checked={checked}
          onChange={this.onChange}
          onClick={this.onClick}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          disabled={disabled || isLoading}
        />
        {(!chevron || isLoading) && (
          <RadioIcon isChecked={checked} isLoading={isLoading} isDisabled={disabled} />
        )}
      </Fragment>
    )

    return (
      <StyledItemRadio
        className={cc(['kirk-item-radio', className, { focus: this.state.focus }])}
        leftTitle={labelTitle}
        leftBody={label}
        rightTitle={data}
        rightTitleDisplay={TextDisplayType.SUBHEADERSTRONG}
        rightBody={dataInfo}
        /* No a11y issue here
            - The input is well wrapped with the label
            - The linter can't access the complex components implementation
          */
        // eslint-disable-next-line jsx-a11y/label-has-associated-control
        tag={<label />}
        leftAddon={leftAddon}
        rightAddon={radio}
        chevron={chevron && !isLoading}
        highlighted={highlighted}
        isClickable={!disabled}
        disabled={disabled}
        hasHorizontalSpacing={hasHorizontalSpacing}
        {...a11yAttrs}
      />
    )
  }
}
