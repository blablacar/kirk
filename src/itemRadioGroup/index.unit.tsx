import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'

import ItemRadio from 'itemRadio'

import ItemRadioGroup, { ItemRadioGroupProps } from './index'

describe('ItemRadioGroup', () => {
  const defaultProps: ItemRadioGroupProps = {
    name: 'name',
    value: 2,
    onChange() {},
    status: ItemRadio.STATUS.DEFAULT,
    children: [
      <ItemRadio label="1" value={1} />,
      <ItemRadio label="2" value={2} />,
      <ItemRadio label="3" value={3} />,
    ],
  }
  it('Should map its children and render them with specific props', () => {
    const itemRadioGroup = renderer
      .create(<ItemRadioGroup {...defaultProps} />)
      .toJSON()
    expect(itemRadioGroup).toMatchSnapshot()
  })
  it('Should listen to all children ItemRadio changes', () => {
    const itemRadioGroup = shallow(<ItemRadioGroup {...defaultProps} />)
    const onChangeListener = itemRadioGroup.instance().onChange
    expect(itemRadioGroup.find({ value: 1 }).prop('onChange')).toEqual(onChangeListener)
    expect(itemRadioGroup.find({ value: 2 }).prop('onChange')).toEqual(onChangeListener)
    expect(itemRadioGroup.find({ value: 3 }).prop('onChange')).toEqual(onChangeListener)
  })
  it('Should change the value in the state when an ItemRadio triggers a change', () => {
    const itemRadioGroup = shallow(<ItemRadioGroup {...defaultProps} />)
    itemRadioGroup.instance().onChange({ name: 'name', value: 'foo' })
    expect(itemRadioGroup.state('value')).toBe('foo')
  })
  it('Should call the onChange prop when an ItemRadio triggers a change', () => {
    const onChangeMock = jest.fn()
    const itemRadioGroup = shallow(
      <ItemRadioGroup {...defaultProps} onChange={onChangeMock} />,
    )
    itemRadioGroup.instance().onChange({ name: 'name', value: 'foo' })
    expect(onChangeMock).toHaveBeenCalledWith({ name: defaultProps.name, value: 'foo' })
  })
})
