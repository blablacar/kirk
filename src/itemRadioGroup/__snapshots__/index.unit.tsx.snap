// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemRadioGroup Should map its children and render them with specific props 1`] = `
.c1 {
  margin: 0;
  font-weight: 400;
}

.c1.kirk-text-button {
  color: #054752;
  font-size: 16px;
  line-height: 20px;
}

.c1.kirk-text-body,
.c1.kirk-text-bodyStrong {
  color: #708C91;
  font-size: 16px;
  line-height: 20px;
}

.c1.kirk-text-caption {
  color: #708C91;
  font-size: 13px;
  line-height: 16px;
}

.c1.kirk-text-title,
.c1.kirk-text-titleStrong {
  color: #054752;
  font-size: 18px;
  line-height: 20px;
}

.c1.kirk-text-display1 {
  color: #054752;
  font-size: 82px;
  line-height: 82px;
}

.c1.kirk-text-display2 {
  color: #054752;
  font-size: 30px;
  line-height: 1.06;
}

.c1.kirk-text-subheader,
.c1.kirk-text-subheaderStrong {
  color: #054752;
  font-size: 22px;
  line-height: 24px;
}

.c1.kirk-text-title,
.c1.kirk-text-titleStrong {
  color: #054752;
  font-size: 18px;
  line-height: 20px;
}

.c1.kirk-text-bodyStrong,
.c1.kirk-text-subheaderStrong,
.c1.kirk-text-titleStrong {
  font-weight: 500;
}

.c3 {
  display: inline-block;
  position: relative;
}

.c3 .kirk-icon-badge {
  position: absolute;
  bottom: 12px;
  left: 12px;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  padding-top: 16px;
  padding-bottom: 16px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border: 0;
  background: none;
}

.c0.kirk-item--clickable:hover {
  background: #EDEDED;
}

a.c0 {
  background: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-tap-highlight-color: rgba(221,221,221,.4);
  -webkit-touch-callout: none;
}

button.c0 {
  cursor: pointer;
  text-align: left;
  width: 100%;
  font-family: inherit;
  outline: none;
  background-color: transparent;
  -webkit-tap-highlight-color: rgba(221,221,221,.4);
  -webkit-touch-callout: none;
}

.c0:disabled {
  cursor: default;
}

.c0 .kirk-item-leftText {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 .kirk-item-rightText {
  margin-left: 16px;
  text-align: right;
}

.c0 .kirk-item-title--withBody {
  margin-bottom: 4px;
}

.c0 .kirk-item-leftAddon,
.c0 .kirk-item-rightAddon {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  min-width: 24px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .kirk-item-leftAddon {
  margin-right: 16px;
}

.c0 .kirk-item-rightAddon {
  margin-left: 16px;
}

.c0.kirk-item--highlighted .kirk-text-title {
  color: #00AFF5;
}

.c2.spinning {
  -webkit-animation: rotator 1.4s linear infinite;
  animation: rotator 1.4s linear infinite;
}

.c2 circle {
  stroke-width: 6;
  stroke-linecap: round;
}

.c2.spinning circle {
  stroke-dasharray: 187;
  stroke-dashoffset: 0;
  -webkit-transform-origin: center;
  -ms-transform-origin: center;
  transform-origin: center;
  -webkit-animation: dash 1.4s ease-in-out infinite;
  animation: dash 1.4s ease-in-out infinite;
}

.c2.thin circle {
  stroke-width: 3;
}

.c2.absolute {
  position: absolute;
}

.c2 circle.inner {
  stroke-width: 0;
}

<ul
  className="jsx-1825249063 kirk-items-list"
  role="radiogroup"
>
  <li
    className="jsx-1825249063 kirk-items-list-item"
  >
    <label
      className="kirk-item kirk-item--clickable kirk-item-radio c0"
    >
      <div
        className="kirk-item-leftText"
      >
        <div
          className="kirk-text kirk-text-body kirk-item-body c1"
        >
          1
        </div>
      </div>
      <div
        className="kirk-item-rightAddon"
      >
        <input
          checked={false}
          disabled={false}
          name="name"
          onChange={[Function]}
          type="radio"
          value={1}
        />
        <svg
          aria-hidden={true}
          className="kirk-icon kirk-icon-circle c2 c3"
          fill="#00AFF5"
          height={24}
          viewBox="0 0 66 66"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          
          <circle
            cx="33"
            cy="33"
            fill="none"
            r="30"
            stroke="#00AFF5"
          />
        </svg>
      </div>
    </label>
  </li>
  <li
    className="jsx-1825249063 kirk-items-list-item"
  >
    <label
      className="kirk-item kirk-item--clickable kirk-item-radio c0"
    >
      <div
        className="kirk-item-leftText"
      >
        <div
          className="kirk-text kirk-text-body kirk-item-body c1"
        >
          2
        </div>
      </div>
      <div
        className="kirk-item-rightAddon"
      >
        <input
          checked={true}
          disabled={false}
          name="name"
          onChange={[Function]}
          type="radio"
          value={2}
        />
        <svg
          aria-hidden={true}
          className="kirk-icon kirk-icon-circle c2 c3"
          fill="#00AFF5"
          height={24}
          viewBox="0 0 66 66"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          
          <circle
            cx="33"
            cy="33"
            fill="none"
            r="30"
            stroke="#00AFF5"
          />
          <circle
            className="inner"
            cx="33"
            cy="33"
            fill="#00AFF5"
            r="18"
          />
        </svg>
      </div>
    </label>
  </li>
  <li
    className="jsx-1825249063 kirk-items-list-item"
  >
    <label
      className="kirk-item kirk-item--clickable kirk-item-radio c0"
    >
      <div
        className="kirk-item-leftText"
      >
        <div
          className="kirk-text kirk-text-body kirk-item-body c1"
        >
          3
        </div>
      </div>
      <div
        className="kirk-item-rightAddon"
      >
        <input
          checked={false}
          disabled={false}
          name="name"
          onChange={[Function]}
          type="radio"
          value={3}
        />
        <svg
          aria-hidden={true}
          className="kirk-icon kirk-icon-circle c2 c3"
          fill="#00AFF5"
          height={24}
          viewBox="0 0 66 66"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          
          <circle
            cx="33"
            cy="33"
            fill="none"
            r="30"
            stroke="#00AFF5"
          />
        </svg>
      </div>
    </label>
  </li>
</ul>
`;
